File: src/main/java/ua/softgroup/matrix/server/supervisor/producer/config/GenericExceptionMapper.java
package ua.softgroup.matrix.server.supervisor.producer.config;

import com.nimbusds.jose.JOSEException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import ua.softgroup.matrix.server.supervisor.producer.exception.JwtException;
import ua.softgroup.matrix.server.supervisor.producer.json.ErrorJson;

import javax.ws.rs.WebApplicationException;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.ext.ExceptionMapper;
import javax.ws.rs.ext.Provider;

/**
 * @author Oleksandr Tyshkovets <sg.olexander@gmail.com>
 */
@Provider
public class GenericExceptionMapper implements ExceptionMapper<Throwable> {

    private static final Logger LOG = LoggerFactory.getLogger(GenericExceptionMapper.class);

    @Override
    public Response toResponse(Throwable throwable) {
        LOG.warn("Jersey module exception: ", throwable);
        return Response.status(getStatusType(throwable).getStatusCode())
                .entity(new ErrorJson(throwable.getLocalizedMessage()))
                .type(MediaType.APPLICATION_JSON)
                .build();
    }

    private Response.StatusType getStatusType(Throwable ex) {
        return (ex instanceof WebApplicationException)
                ? ((WebApplicationException) ex).getResponse().getStatusInfo()
                : (ex instanceof JOSEException || ex instanceof JwtException
                        ? Response.Status.FORBIDDEN
                        : Response.Status.INTERNAL_SERVER_ERROR);
    }
}

File: src/main/scala/ua/softgroup/matrix/server/supervisor/producer/config/GenericExceptionMapper.scala
package ua.softgroup.matrix.server.supervisor.producer.config

import javax.ws.rs.{NotAuthorizedException, NotFoundException, WebApplicationException}
import javax.ws.rs.core.{MediaType, Response}
import javax.ws.rs.ext.{ExceptionMapper, Provider}

import com.nimbusds.jose.JOSEException
import org.slf4j.LoggerFactory
import ua.softgroup.matrix.server.supervisor.producer.exception.JwtException
import ua.softgroup.matrix.server.supervisor.producer.json.v2.ErrorJson

/**
  * @author Oleksandr Tyshkovets <sg.olexander@gmail.com> 
  */
@Provider
class GenericExceptionMapper extends ExceptionMapper[Throwable] {

  private val logger = LoggerFactory.getLogger(this.getClass)

  override def toResponse(throwable: Throwable): Response = {
    if (!(throwable.isInstanceOf[NotFoundException] || throwable.isInstanceOf[NotAuthorizedException]))
      logger.warn("Jersey module exception: ", throwable)

    Response
      .status(getStatusType(throwable).getStatusCode)
      .entity(new ErrorJson(throwable.getLocalizedMessage))
      .`type`(MediaType.APPLICATION_JSON)
      .build
  }

  private def getStatusType(ex: Throwable) = {
    ex match {
      case e: WebApplicationException => e.getResponse.getStatusInfo
      case _: JOSEException | _: JwtException => Response.Status.FORBIDDEN
      case _ => Response.Status.INTERNAL_SERVER_ERROR
    }
  }

}
